### DISCUSSION TITLE
Sharing data between Dialog and Program.cs

### DISCUSSION THREAD

**Discussion by kplatter** [oleg-shilo/wixsharp#1469] [General]
How can I share data that is collected on a custom Dialog and use it as part of a CustomAction?  

I have found that I can share data between Diaglogs using ((MsiRuntime)shell?.RuntimeContext).Data["variable"] but so far have not found how I can share it and use it in a CustomAction...  I am thinking I need to use a session.Property somehow but not sure how...

**Comment by oleg-shilo** [comment]
You have a few options here.

1. You can share the data by pushing it directly to the session:
    ```C#
    // write
    session["my-prop"] = "my-value";

    // read
    var my_value = session.Property("my-prop");
    ```
    It's simple but your session will be already disposed if you access it at the end of the installation (e.g. accessing from the deferred actions). Thus you need to take special care about that specific property so it's preserved after install: 
    ```C#
    project.DefaultDeferredProperties += ",my-prop";
    ```  
2. You can use `args.Data` as it is nothing else but a syntactic sugar for accessing `session["WIXSHARP_RUNTIME_DATA"]` property containing a serializable dictionary. This property is already channelled to the end of the session.
    - You can either convert session to `SetupEventArgs` with an easy access to the `Data[...]` member:
        ```C#
        SetupEventArgs args = session.ToEventArgs();
        ```
    - Or you can simply drop the use of CA and instead use project events. They are nothing else but wrappers around classic CA. And evens already have easy access to the `args.Data[...]` so there is no need for any conversion. This is my preferred approach.

**Comment by kplatter** [comment]
How do you access the session in the Dialog?  When I try to use (shell?.RuntimeContext as MsiRuntime).Session it does not seem to be the same session or work as I would expect... (If it makes any difference I am using WIX4)

session is not available to do session["my-prop"] = "my-value"; in the dialog???  It is a WPF dialog if that matters...

I am trying to save the data in the Dialog GoNext_Click event

``` C#
       private async void GoNext_Click(object sender, RoutedEventArgs e)
       {
                (shell?.RuntimeContext as MsiRuntime).MsiSession.CustomActionData["TEST1"] = cmbServerName.Text;

                (shell?.RuntimeContext as MsiRuntime).Session["TEST2"] = cmbServerName.Text;

                Host?.MsiRuntime().Data.Add("TEST3", "THIS IS TEST3");
       }
```

and access the data in a ElevatedManagedAction once the files have been installed

new ElevatedManagedAction(CustomActions.MyAction, Return.check, When.After, Step.InstallFiles, Condition.NOT_Installed),

**Comment by kplatter** [comment]
I was able to access the data from the Dialog in AfterInstall by using the following
``` C#
//Dialog
((MsiRuntime)shell?.RuntimeContext).Data.Add("SQLSERVER", cmbServerName.Text);

//AfterInstall
string server = e.Data["SQLSERVER"];
```

But still have not figured out how to access it in CustomAction, as you said AfterInstall should work for what I need but would be good to know how to access it in CA

**Comment by oleg-shilo** [comment]
> But still have not figured out how to access it in CustomAction,...

I provided the solution in my last post. You need that `ToEventArgs` conversion:

```C#
[CustomAction]
public static ActionResult MyCA(Session session)
{
        var value = session.ToEventArgs().Data["SQLSERVER"];
        . . .
```

**Comment by kplatter** [comment]
I tried that and got 

'Session' does not contain a definition for 'ToEventArgs' and no accessible extension method 'ToEventArgs' accepting a first argument of type 'Session' could be found (are you missing a using directive or an assembly reference?).

This error occurs when you try to call a method or access a class member that does not exist.

Am I missing something?

**Comment by oleg-shilo** [comment]
Are you using WiX4 or WiX3 stream?

**Comment by kplatter** [comment]
WIX4

**Comment by oleg-shilo** [comment]
Ensure you are using the latest release (nuget package). This method was added recently. Quite possibly you are using the older version.

![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/520ec437-6419-4343-80c3-78d8899257fc)

**Comment by kplatter** [comment]
That did the trick, for some reason (and I just installed wixsharp a couple days ago) I was running 2.0.0

**Comment by kplatter** [comment]
The session to event args is now giving me 

Info: Cannot access session details from a non-immediate custom action
Info: Exception thrown by custom action:
Info: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> WixToolset.Dtf.WindowsInstaller.InstallerException: Cannot access session details from a non-immediate custom action

from this

new ElevatedManagedAction(CustomActions.MyCustomAction, Return.check, When.After, Step.InstallFiles, Condition.NOT_Installed),

**Comment by oleg-shilo** [comment]
The template (as all templates) points to the latest version of the package at the time of the template release. But the package evolves independently. So it's always a good idea to start with the package update.

Unless you did it just before the package release. Also about 2 days ago. :)

BTW that conversion is nothing magic. One can implement it very easily. IE WiX3 stream is yet to receive this conversion extension:

```C#
static class MyExtensions
{
    public static SetupEventArgs ToEventArgs(this Session session)
    {
        ManagedProject.Init(session);
        return ManagedProject.Convert(session);
    }
}
```

**Comment by kplatter** [comment]
I was able to access the method, but get an error when it runs...

The session to event args is now giving me

Info: Cannot access session details from a non-immediate custom action
Info: Exception thrown by custom action:
Info: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> WixToolset.Dtf.WindowsInstaller.InstallerException: Cannot access session details from a non-immediate custom action

from this

new ElevatedManagedAction(CustomActions.MyCustomAction, Return.check, When.After, Step.InstallFiles, Condition.NOT_Installed),

**Comment by oleg-shilo** [comment]
I will have a look at it. I probably mean that the conversion touches some non-channelled properties as you are accessing it from the deferred action.

> new ElevatedManagedAction(CustomActions.MyCustomAction, Return.check, When.After, Step.InstallFiles, Condition.NOT_Installed),

Then you are effectively setting up an elevated ManagedActin that is already done for you. You are using an equivalent of `project.AfterInstall`.
Not sure, what advantage it gives you comparing to the event.

**Comment by oleg-shilo** [comment]
Have a look here: https://github.com/oleg-shilo/wixsharp/wiki/Managed-Setup-Model#event-driven-custom-actions

**Comment by kplatter** [comment]
Thanks, I will move to using the AfterInstall but if you have any luck in getting ToEventArgs working let us know :)

**Reply by oleg-shilo** [reply]
The problem with accessing `SetupEventArgs.AppData` is that in your case it is accessed from the deferred action and for all raw elevated actions you need to specify explicitly that you need this data to be available after the session is terminated. Tus you need to set `UsesProperties`:

```C#
new ElevatedManagedAction(CustomActions.MyCustomAction, Return.check, When.After, Step.InstallFiles, Condition.NOT_Installed)
{
    UsesProperties = "WIXSHARP_RUNTIME_DATA"
}
```
When accessing the data I found that ToEvenArgs will throw for the very same reason being called from deferred action. Thus you need to access only the preserved part of the session data:

```C#
[CustomAction]
public static ActionResult MyCustomAction(Session session)
{
    Debug.Assert(false);
    try
    {
        var data = new AppData().InitFrom(session.Property("WIXSHARP_RUNTIME_DATA"));
        var value = data["SQLSERVER"];
        . . .
```

Now you probably see why I was saying that events are way more convenient.

**Comment by kplatter** [comment]
Ty that did the trick :)
